File: Threema/SupportViewController.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2012-2022 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "SupportViewController.h"
#import "MyIdentityStore.h"
#import "Utils.h"
#import <MBProgressHUD/MBProgressHUD.h>
#import "BundleUtil.h"

@interface SupportViewController ()

@end

@implementation SupportViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.webView = [[WKWebView alloc] initWithFrame:self.view.frame];
    self.webView.allowsLinkPreview = NO;
    self.webView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.webView.navigationDelegate = self;
    [self.view addSubview:self.webView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    NSURL *supportUrl = nil;
    
    if ([MyIdentityStore sharedMyIdentityStore].licenseSupportUrl.length > 0) {
        supportUrl = [NSURL URLWithString:[MyIdentityStore sharedMyIdentityStore].licenseSupportUrl];
    }
    
    if (supportUrl == nil) {
        NSMutableArray *queryItems = [[NSMutableArray alloc] init];
        [queryItems addObject:[NSURLQueryItem queryItemWithName:@"lang" value:[[[NSBundle mainBundle] preferredLocalizations] objectAtIndex:0]]];
        [queryItems addObject:[NSURLQueryItem queryItemWithName:@"version" value:[Utils getClientVersion]]];
        [queryItems addObject:[NSURLQueryItem queryItemWithName:@"identity" value:[MyIdentityStore sharedMyIdentityStore].identity]];
        
        NSURLComponents *components = [NSURLComponents componentsWithString:[BundleUtil objectForInfoDictionaryKey:@"ThreemaSupportURL"]];
        components.queryItems = queryItems;
        supportUrl = components.URL;
    }
    
    [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    [self.webView loadRequest:[NSURLRequest requestWithURL:supportUrl cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:15]];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    [MBProgressHUD hideHUDForView:self.view animated:YES];    
}

- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler {
    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
        if (navigationAction.request.URL != nil) {
            [[UIApplication sharedApplication] openURL:navigationAction.request.URL options:@{} completionHandler:nil];
            decisionHandler(WKNavigationActionPolicyCancel);
            return;
        }
    }
    decisionHandler(WKNavigationActionPolicyAllow);
}

- (BOOL)shouldAutorotate {
    return YES;
}

-(UIInterfaceOrientationMask)supportedInterfaceOrientations {
    if (SYSTEM_IS_IPAD) {
        return UIInterfaceOrientationMaskAll;
    }
    
    return UIInterfaceOrientationMaskAllButUpsideDown;
}

@end

File: Threema/Settings/SupportViewController.swift
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2021-2024 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import CocoaLumberjackSwift
import Foundation
import MBProgressHUD
import SafariServices
import WebKit

class SupportViewController: UIViewController {
    
    private var webView: WKWebView!
    private var myIdentity = BusinessInjector().myIdentityStore
    
    // MARK: - Lifecycle
    
    override func loadView() {
        let webConfiguration = WKWebViewConfiguration()
        webView = WKWebView(frame: .zero, configuration: webConfiguration)
        webView.navigationDelegate = self
        view = webView
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(themeChanged),
            name: Notification.Name(kNotificationColorThemeChanged),
            object: nil
        )

        // Set correct navigation bar edge appearance
        themeChanged()
        
        if let myURL = createURL() {
            let myRequest = URLRequest(url: myURL)
            webView.load(myRequest)
            MBProgressHUD.showAdded(to: webView, animated: true)
        }
        else {
            DDLogError("Url could neither be found nor created")
            searchContact()
        }
    }
    
    // MARK: - Helper methods
    
    func createURL() -> URL? {
        if let licenseURL = myIdentity.licenseSupportURL, let url = URL(string: licenseURL), !licenseURL.isEmpty {
            let supportURL = url
            return supportURL
        }
        else {
            var queryItems = [URLQueryItem]()
            queryItems.append(URLQueryItem(name: "lang", value: Bundle.main.preferredLocalizations[0]))
            queryItems.append(URLQueryItem(name: "version", value: ThreemaUtility.clientVersion))
            queryItems.append(URLQueryItem(name: "identity", value: myIdentity.identity))
            queryItems.append(URLQueryItem(name: "theme", value: Colors.theme == .dark ? "dark" : "light"))
            guard let urlComp = BundleUtil.object(forInfoDictionaryKey: "ThreemaSupportURL") as? String else {
                return nil
            }
            var components = URLComponents(string: urlComp)
            components?.queryItems = queryItems
            return components?.url
        }
    }
    
    func searchContact() {
        ContactStore.shared()
            .addContact(
                with: "*SUPPORT",
                verificationLevel: Int32(kVerificationLevelUnverified),
                onCompletion: { contact, _ in
                    self.moveToSupportChat(contact: contact!)
                },
                onError: { error in
                    DDLogError("\(error)")
                }
            )
    }
    
    func moveToSupportChat(contact: ContactEntity) {
        let message = "My app version: \(ThreemaUtility.clientVersionWithMDM)"
        let info = [
            kKeyContact: contact,
            kKeyForceCompose: NSNumber(value: true),
            kKeyText: message,
        ] as [String: Any]
        
        DispatchQueue.main.async {
            NotificationCenter.default.post(
                name: NSNotification.Name(rawValue: kNotificationShowConversation),
                object: nil,
                userInfo: info
            )
        }
    }
    
    func showWebsite(_ url: URL) {
        let config = SFSafariViewController.Configuration()
        let safariView = SFSafariViewController(url: url, configuration: config)
        present(safariView, animated: true)
    }
    
    @objc private func themeChanged() {
        navigationItem.scrollEdgeAppearance = Colors.theme == .dark ? Colors.defaultNavigationBarAppearance() : Colors
            .transparentNavigationBarAppearance()
    }
}

// MARK: - WKNavigationDelegate

extension SupportViewController: WKNavigationDelegate {
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        MBProgressHUD.hide(for: webView, animated: true)
    }
    
    func webView(
        _ webView: WKWebView,
        decidePolicyFor navigationAction: WKNavigationAction,
        decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
    ) {
        
        if let url = navigationAction.request.url, navigationAction.navigationType == .linkActivated {
            if let scheme = url.scheme, scheme.hasPrefix("http") || scheme.hasPrefix("https") {
                showWebsite(url)
            }
            else {
                searchContact()
            }
            decisionHandler(.cancel)
            return
        }
        decisionHandler(.allow)
    }
}
